### PR TITLE
nixos/repart-image: pass partition attrs to builder instead of JSON file

### PR DESCRIPTION
**Pull request by WilliButz**
## Description of changes

Having access to the original Nix partition definitions in the builder should make it a bit easier to manipulate them and still provide access to the manipulated results.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by WilliButz** [comment]
I tried to go with `passAsFile` for `partitionsJSON` instead, but that seems to work differently with structuredAttrs enabled(?)

**Comment by WilliButz** [comment]
@ofborg test appliance-repart-image-verity-store

**Comment by RaitoBezarius** [comment]
While we are it, can we sed also the TBD in the repart-output.json by the actual roothash?

**Comment by WilliButz** [comment]
@ofborg test appliance-repart-image-verity-store

File: nixos/modules/image/repart-verity-store.nix [hunk 1]
```diff
@@ -194,6 +193,24 @@ in
                 chmod +w ${config.image.repart.imageFileBasename}.raw
               '';
 
+              # replace "TBD" with the original roothash values
+              preInstall = ''
```
Comment (nikstur) [inline nixos/modules/image/repart-verity-store.nix:197]: Shouldn't the second derivation already return a repart-output.json that contains the final values?

Comment (WilliButz) [reply nixos/modules/image/repart-verity-store.nix:197]: The output from the intermediate image includes the hashes
```
<snip>
  "activity" : "create",
  "roothash" : "7c915215f4..............."
}
</snip>
```
but in the final image there is just the literal "TBD" string. I don't think repart tries to retrieve and match hashes for verity siblings again when the partitions are already pre-populated.
Output from the build of `finalImage` looks like this:
```
<snip>
  "activity" : "unchanged",
  "roothash" : "TBD"
}
</snip>
```

**Review by nikstur — APPROVED** [review]

**Review by RaitoBezarius — APPROVED** [review]

**Pull request merged**